### PR TITLE
Various cleanups in the Inliner

### PR DESCRIPTION
**Pull request by lrytz**

### CODE REVIEW

**Comment by lrytz** [comment]
review by @retronym

**Comment by sjrd** [comment]
The last commit (callsite `@inline/@noinline`) doesn't sound like a "cleanup". It's an entirely new feature.

It also has consequences that go beyond the inliner itself. First, other back-ends could reuse the attachments to also take callsite `@inline/@noinline` into account (I've had a mind to do this in Scala.js for some time).

But more importantly, how is this going to interact with sbt's incremental compilation? Currently, if you change the body of an `@inline` method, sbt considers it a binary incompatible change to the method (even if the interface is untouched) and will recompile callsites. Now, if you don't have `@inline` at definition site, but you have `@inline` at callsite, how would sbt know to recompile the client when the callee's body is updated?

**Comment by lrytz** [comment]
Thanks @sjrd, I was sloppy in the PR title - we should indeed discuss this a bit more.

In terms of interaction with the incremental compiler: there's noting new. Incremental compilation in sbt is incorrect when the optimizer is enabled, that is the case in 2.11.7, and we're not planning to fix this. Example:

```
➜  sbtproj  ls src/main/scala/
A.scala    B.scala    Main.scala
➜  sbtproj  cat src/main/scala/*
class A {
  @inline final def f = 3
}
class B {
  @inline final def f = (new A).f
}
object Main extends App {
  println((new B).f)
}
➜  sbtproj  cat build.sbt
scalaVersion := "2.11.7"

scalacOptions += "-optimize"
➜  sbtproj  sbt clean run
[info] Compiling 3 Scala sources to /Users/luc/tmp/sbtproj/target/scala-2.11/classes...
[info] Running Main
3
```

Now I edit `A.scala`

```
➜  sbtproj  cat src/main/scala/A.scala
class A {
  @inline final def f = 3333
}
➜  sbtproj  sbt run
[info] Compiling 1 Scala source to /Users/luc/tmp/sbtproj/target/scala-2.11/classes...
[info] Running Main
3
➜  sbtproj  sbt clean run
[info] Compiling 3 Scala sources to /Users/luc/tmp/sbtproj/target/scala-2.11/classes...
[info] Running Main
3333
```

**Comment by sjrd** [comment]
> Incremental compilation in sbt is incorrect when the optimizer is enabled, that is the case in 2.11.7, and we're not planning to fix this.

Oh, OK then.

File: src/compiler/scala/tools/nsc/backend/jvm/opt/Inliner.scala [hunk 1]
```diff
@@ -48,19 +47,7 @@ class Inliner[BT <: BTypes](val btypes: BT) {
       // need to run DCE before inlining g.
       eliminateUnreachableCodeAndUpdateCallGraph(callee.callee, callee.calleeDeclarationClass.internalName)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/Inliner.scala]: todo: move this down to `def inline`?

**Comment by lrytz** [comment]
@sjrd i think the `@inline`-ascription should work for scala.js - anything you'd like to have differently from your side? Here's how I check the AST if an Apply is ascribed (the tree attachment is added during Typers): https://github.com/lrytz/scala/commit/6f72cb19b266c26e154df75af444983ca41865ec#diff-2e0fd5042aa83407a80a7557cc2a666bR710

Test cases here: https://github.com/lrytz/scala/commit/6f72cb19b266c26e154df75af444983ca41865ec#diff-c108d2950579fe9645aa47c8a1982256R1059

**Comment by sjrd** [comment]
It looks like I could do a similar test in Scala.js, yes. I don't think it should be any different. :-)

**Comment by lrytz** [comment]
green now - review by @retronym

**Comment by retronym** [comment]
LGTM!

**Pull request merged**